### PR TITLE
SDIT-2336 Resynchronise alerts

### PR DESCRIPTION
**Pull request by andymarke**
deletes alerts on active booking and replace them new alerts

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt [hunk 1]
```diff
@@ -32,11 +32,10 @@ class AlertsService(
   val alertTypeRepository: ReferenceCodeRepository<AlertType>,
   val workFlowActionRepository: ReferenceCodeRepository<WorkFlowAction>,
 ) {
-  fun getAlert(bookingId: Long, alertSequence: Long): AlertResponse =
-    offenderBookingRepository.findByIdOrNull(bookingId)?.let { booking ->
-      offenderAlertRepository.findById_OffenderBookingAndId_Sequence(booking, alertSequence)?.toAlertResponse()
-        ?: throw NotFoundException("Prisoner alert not found for alertSequence=$alertSequence")
-    } ?: throw NotFoundException("Prisoner booking not found for bookingId=$bookingId")
+  fun getAlert(bookingId: Long, alertSequence: Long): AlertResponse = offenderBookingRepository.findByIdOrNull(bookingId)?.let { booking ->
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt:35]: formatting

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt [hunk 2]
```diff
@@ -133,28 +173,24 @@ class AlertsService(
...
-        return PrisonerAlertsResponse(
-          latestBookingAlerts = latestBooking.alerts.map { it.toAlertResponse() }.sortedBy { it.alertSequence },
-        )
-      } ?: throw NotFoundException("Prisoner with offender $offenderNo not found with any bookings")
-
-  fun getActiveAlertsForReconciliation(offenderNo: String): PrisonerAlertsResponse =
-    PrisonerAlertsResponse(
-      latestBookingAlerts = getAlerts(offenderNo).latestBookingAlerts.filter { it.isActive },
-    )
+  fun getAlerts(offenderNo: String): PrisonerAlertsResponse = offenderBookingRepository.findAllByOffenderNomsId(offenderNo).takeIf { it.isNotEmpty() }
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt:176]: formatting only

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt [hunk 3]
```diff
@@ -133,28 +173,24 @@ class AlertsService(
...
+      return PrisonerAlertsResponse(
+        latestBookingAlerts = latestBooking.alerts.map { it.toAlertResponse() }.sortedBy { it.alertSequence },
+      )
+    } ?: throw NotFoundException("Prisoner with offender $offenderNo not found with any bookings")
+
+  fun getActiveAlertsForReconciliation(offenderNo: String): PrisonerAlertsResponse = PrisonerAlertsResponse(
+    latestBookingAlerts = getAlerts(offenderNo).latestBookingAlerts.filter { it.isActive },
+  )
 
-  fun getAlerts(bookingId: Long): BookingAlertsResponse =
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt:150]: formatting only

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt [hunk 4]
```diff
@@ -133,28 +173,24 @@ class AlertsService(
...
+    ?.let { booking -> offenderAlertRepository.findAllById_OffenderBooking(booking).map { it.toAlertResponse() } }
+    ?.let { BookingAlertsResponse(it) }
+    ?: throw NotFoundException("Prisoner with booking $bookingId not found")
 }
 
 private fun Staff?.asDisplayName(): String? = this?.let { "${it.firstName} ${it.lastName}" }
 
-private fun OffenderBooking.hasActiveAlertOfCode(alertCode: String) =
-  this.alerts.firstOrNull { it.alertCode.code == alertCode && it.alertStatus == ACTIVE } != null
+private fun OffenderBooking.hasActiveAlertOfCode(alertCode: String) = this.alerts.firstOrNull { it.alertCode.code == alertCode && it.alertStatus == ACTIVE } != null
```
Comment (andymarke) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt:196]: formatting only

File: src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt [hunk 5]
```diff
@@ -77,6 +76,47 @@ class AlertsService(
...
+          alertType = alertTypeRepository.findByIdOrNull(AlertType.pk(alertCode.parentCode!!))!!,
+          alertCode = alertCode,
+          authorizePersonText = request.authorisedBy,
+          alertStatus = if (request.isActive) ACTIVE else INACTIVE,
+          verifiedFlag = false,
+          commentText = request.comment,
+          createUsername = request.createUsername,
+        ).also {
+          it.addWorkFlowLog(workActionCode = workActionCode)
+        },
```
Comment (mikehalmamoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/alerts/AlertsService.kt:106]: I guess this means you'd lose any old work flow logs?

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**